<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px"> 

## Lab: SQL Practice

_Author: Matt Brems_

### Required Preparation

In order to run SQL queries within Python/this Jupyter notebook, you will need to install the following:

```bash
conda install sqlalchemy
conda install psycopg2
```

If you get an error trying to install `psycopg2`, you can install a stand-alone package (sufficient for everything you'll need to complete this lab) by entering the following anywhere in the terminal: 

```bash 
pip install psycopg2-binary
```

(If `pip` doesn't work for you, try `pip3 install psycopg2-binary`.)

### Instructions

In this lab, you're going to have ten prompts. Each prompt will require you to do two things:
1. Write a SQL query.
2. Use the result of that query to answer the question.

The purpose of this lab is to get you to practice your SQL - an **important** skill in data science! While it is possible to get these answers using Pandas, it should be your goal to do everything only in SQL.

**Data**: The data used in this lab is the Iowa liquor database. It contains various tables of transactions. The schema for this database is here:

<img src="./images/schema.png" alt="schema" width="750"/>

In [4]:
# Import pandas and create_engine from `sqlalchemy`
import pandas as pd
from sqlalchemy import create_engine
# Set up the engine to access the data
engine = \
create_engine('postgresql://analytics_student:analyticsga@analyticsga-psql.generalassemb.ly:5432/iowa_liquor_sales_database')

### Prompt 1

How many rows are there in the `stores` table of the Iowa liquor database?

In [4]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT *
FROM stores
"""

pd.read_sql_query(sql, engine).shape

(1973, 5)

**Prompt 1 Answer**: 1973 rows

### Prompt 2

If you sort the stores in alphabetical order by name, which store is fourth?

In [11]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT name
FROM stores
GROUP BY name
ORDER BY name ASC
"""
pd.read_sql_query(sql, engine).head()

name
0  218 Fuel Express & Chubby's Liquor
1          3047 Cub Foods / Iowa City
2               3060 Cub Foods / Ames
3         3061 Cub Foods / Sioux City
4                  3rd St Convenience

**Prompt 2 Answer**: 3061 Cub Foods / Sioux City

### Prompt 3

Among those sales whose category name is `IMPORTED VODKA`, if you sort by vendor in alphabetical order, what is the bottle price and number of bottles bought in the first 3 transactions?
> You should give us three sets of numbers - one for each transaction.

In [13]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT category_name, btl_price, bottle_qty
FROM sales
WHERE category_name LIKE '%%IMPORTED VODKA%%'
ORDER BY vendor ASC
"""

pd.read_sql_query(sql, engine).head()

category_name btl_price  bottle_qty
0  IMPORTED VODKA     $8.99          12
1  IMPORTED VODKA    $28.50          30
2  IMPORTED VODKA    $28.50           6
3  IMPORTED VODKA    $28.50           3
4  IMPORTED VODKA    $28.50           6

**Prompt 3 Answer**: 
1. $ 8.99 / 12 qty

2. $28.50 / 30 qty

3. $28.50 /  6 qty

### Prompt 4

What is the total amount of liquor sold in February 2015?

In [12]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT SUM(sales.bottle_qty)
FROM sales
WHERE date BETWEEN '2015-2-01' AND '2015-2-28'
"""
pd.read_sql_query(sql, engine)

sum
0  1591962

**Prompt 4 Answer**: 1,591,962 bottles

### Prompt 5

Among all transactions where `IMPORTED VODKA` was sold, which vendor has the most transactions? Report the total number of transactions, the number of bottles sold, and the total amount of revenue from these transactions. Rename the columns to make sure there isn't any confusion as to what they mean!

In [14]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT category_name, vendor, SUM(item) as total_transaction, 
       SUM(bottle_qty) as bottle_sold, SUM(total) as total_revenue
FROM sales
WHERE category_name LIKE '%%IMPORTED VODKA%%'
GROUP BY category_name, vendor
ORDER BY total_transaction DESC
"""

pd.read_sql_query(sql, engine).head(3)

category_name                            vendor  total_transaction  \
0         IMPORTED VODKA  Pernod Ricard USA/Austin Nichols         1443447940   
1  IMPORTED VODKA - MISC                   Diageo Americas         1020370680   
2         IMPORTED VODKA               Bacardi U.S.A. Inc.          909997106   

   bottle_sold  total_revenue  
0       403450     7880365.88  
1       129444     2352586.47  
2       236047     5771242.07

**Prompt 5 Answer**: Pernod Ricard USA/Austin Nichols

### Prompt 6

The reason we sell things is to make money. In sales, the term **markup** means the amount of extra money charged for a product over the cost to make that product. (You can think of the markup as the profit for that product.)

In Iowa, the law states the minimum price at which liquor may be sold.

Calculate the markup (name this `Markup`) that shows the difference between the `btl_price` and the `state_btl_cost`. 

> If you haven't subtracted columns together before, [this StackOverflow question](https://stackoverflow.com/questions/7536996/multiplying-two-columns-in-sql-server/7537059) may be helpful.

In [2]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT btl_price, state_btl_cost, (btl_price-state_btl_cost) as Markup
FROM sales
LIMIT 5
"""
pd.read_sql_query(sql, engine).head()

btl_price state_btl_cost markup
0    $17.24         $11.49  $5.75
1    $17.24         $11.49  $5.75
2    $17.24         $11.49  $5.75
3    $25.73         $17.15  $8.58
4     $7.50          $5.00  $2.50

### Prompt 7

Calculate the "Percentage Markup" by dividing markup by the state bottle cost. (Rather than the total profit per bottle, this will tell you how much stores are increasing their price over the state minimum.)

After calculating the "Percentage Markup," calculate the average percentage markup per **store**. Which store has the largest average markup, and what is that average markup?

In [7]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT st.name, AVG((btl_price-state_btl_cost)/state_btl_cost*100) as avg_markup
FROM sales as ss
LEFT JOIN stores as st
ON ss.store = st.store
GROUP BY st.name
"""
pd.read_sql_query(sql, engine)

name  avg_markup
0     218 Fuel Express & Chubby's Liquor   50.130569
1                     3rd St Convenience   50.226254
2                   7 Rayos Liquor Store   50.157033
3                             Abby Lea's   50.261030
4                     Ackley Super Foods   50.207467
...                                  ...         ...
1301              World Liquor & Tobacco   50.185393
1302                  Xo Food And Liquor   50.080319
1303                Zapf's Pronto Market   50.460664
1304                      Z's Quickbreak   50.117824
1305                                None   50.176541

[1306 rows x 2 columns]

**Prompt 7 Answer**: 

### Prompt 8

Building off of your last query, retrieve the five stores with the highest average markup percentage. Along with the store IDs and average markup percentage, please return the minimum and maximum markup percentage for those stores, as well as the names and addresses of those store.

What are the names of the five stores?

In [14]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT st.name AS store_name, AVG((btl_price-state_btl_cost)/state_btl_cost*100) as avg_markup,
       st.store AS ID,st.store_address AS address,
       MIN((btl_price-state_btl_cost)/state_btl_cost*100) as min_markup,
       MAX((btl_price-state_btl_cost)/state_btl_cost*100) as max_markup
FROM sales as ss
LEFT JOIN stores as st
ON ss.store = st.store
GROUP BY st.name,st.store,st.store_address
ORDER BY avg_markup DESC
"""
pd.read_sql_query(sql, engine).head()

store_name  avg_markup      id  \
0      Wal-Mart 0841 / Tipton   52.547124  4013.0   
1  Wal-Mart 1546 / Iowa Falls   52.149880  4024.0   
2  Wal-Mart 1683 / Shenandoah   51.890342  4266.0   
3    Wal-Mart 3394 / Atlantic   51.509378  3833.0   
4   Wal-Mart 2935 / Knoxville   51.459084  3660.0   

                                             address  min_markup  max_markup  
0  1126 Highway 38 North\nTipton, IA 527720000\n(...   49.857143  670.876289  
1  840 S Oak\nIowa Falls, IA 501260000\n(42.50295...   49.846154  670.876289  
2  705 S Fremont\nShenandoah, IA 516010000\n(40.7...   49.846154  670.876289  
3  1905 East 7th St\nAtlantic, IA 500220000\n(41....   49.846154  670.876289  
4  814 W Bell Ave\nKnoxville, IA 501380000\n(41.3...   49.846154  670.876289

**Prompt 8 Answer**: Wal-Mart

### Prompt 9

Which nine counties sell the highest number of bottles of liquor per capita (per person)? Calculate the average bottle size of the bottles of liquor sold by these counties rounded to the nearest first decimal place. Of the nine counties selling the highest number of bottles of liquor per capita, which county sells (on average) the largest bottle size?

**_BONUS FUN_**: This question can (also) be solved with a SQL subquery - that is, a query that queries a query! If you'd like to give this a try, you can read more about it [here](https://www.dofactory.com/sql/subquery).

In [24]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT c.county, (SUM(bottle_qty)/c.population) as liq_per_cap,
       ROUND(AVG(liter_size),1) as avg_size
FROM sales as s
LEFT JOIN counties as c
ON s.county = c.county
GROUP BY c.county
ORDER BY liq_per_cap DESC
"""

pd.read_sql_query(sql, engine).head(10)

county  liq_per_cap  avg_size
0           None          NaN     774.9
1      Dickinson         22.0     974.2
2     Black Hawk         15.0     861.0
3           Polk         15.0     891.9
4    Cerro Gordo         14.0     986.4
5        Johnson         13.0     895.2
6          Scott         13.0     869.9
7           Linn         12.0     885.1
8        Kossuth         11.0    1040.5
9  Pottawattamie         11.0     890.9

**Prompt 9 Answer**: Kossuth

### Prompt 10 (advanced but required!)

In the `sales` table, three columns are `btl_price`, `bottle_qty`, and `total`. The `total` for a transaction _should be_ the product of `btl_price` and `bottle_qty`. How many transactions have a value of `total` that is not equal to `btl_price` time `bottle_qty`?

> There will be a type error as well! We cannot compare type `money` to type `real`. We have to convert `total` to type `money` so that we can compare the two directly. [This link](https://docs.microsoft.com/en-us/sql/t-sql/functions/cast-and-convert-transact-sql?view=sql-server-ver15#syntax) may be helpful to you!

In [6]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT btl_price, bottle_qty, total, (btl_price*bottle_qty)::MONEY AS real_total
FROM sales
WHERE total::MONEY <> (btl_price*bottle_qty)::MONEY
LIMIT 10
"""
pd.read_sql_query(sql, engine)

Empty DataFrame
Columns: [btl_price, bottle_qty, total, real_total]
Index: []

**Prompt 10 Answer**: Null

### MOREEEEE!

In [13]:
#How many total products are in the Products table?
sql=""" 
SELECT COUNT(DISTINCT item_no) as total_product 
FROM products """
pd.read_sql_query(sql, engine)

total_product
0           9977

In [20]:
# Who are the top most diverse vendors (highest number of distinct products)? How many different products do they have?
sql=""" 
SELECT vendor_name, COUNT(DISTINCT item_no) as dist_prod 
FROM products 
GROUP BY vendor_name
ORDER BY dist_prod DESC 
LIMIT 1 """
pd.read_sql_query(sql, engine)

vendor_name  dist_prod
0  Jim Beam Brands        925

In [32]:
# Which products sell the best by total number of unit sales? 
sql="""
SELECT c.county, SUM(pack) AS unit 
FROM sales as s 
LEFT JOIN counties as c
ON s.county = c.county 
GROUP BY c.county 
ORDER BY unit DESC LIMIT 1
"""
pd.read_sql_query(sql, engine)

county     unit
0   Polk  6966015

In [35]:
# Which products sell the best by total dollar value of sales?
sql=""" 
SELECT c.county, s.total 
FROM sales as s LEFT JOIN counties as c
ON s.county = c.county 
GROUP BY c.county, s.total 
ORDER BY total DESC 
LIMIT 1
"""
pd.read_sql_query(sql, engine)


county     total
0   Polk  94590.72

In [40]:
# What are the top 10 categories of liquor sold based on the total amount of sales revenue?
sql=""" 
SELECT category_name, SUM(total) as total 
FROM sales 
GROUP BY category_name, total
ORDER BY total DESC 
LIMIT 10 """
pd.read_sql_query(sql, engine)

category_name       total
0                          SPICED RUM  1987578.00
1                  TENNESSEE WHISKIES  1837478.40
2                     WHISKEY LIQUEUR  1805518.80
3                          SPICED RUM  1717198.08
4                   CANADIAN WHISKIES  1652138.88
5                      IMPORTED VODKA  1360442.88
6                      IMPORTED VODKA  1359814.80
7                      80 PROOF VODKA  1287778.32
8  MISC. AMERICAN CORDIALS & LIQUEURS  1281960.00
9               STRAIGHT RYE WHISKIES  1246703.04

In [10]:
# Which rum products have sales greater than $10,000? How about whiskey or vodka products?
sql=""" 
SELECT (SELECT category_name
        FROM sales
        WHERE category_name LIKE '%%RUM%%' and total > 10000
        ORDER BY total DESC
        LIMIT 1) as RUM,
        (SELECT category_name
        FROM sales
        WHERE category_name LIKE '%%WHISKEY%%' and total > 10000
        ORDER BY total DESC 
        LIMIT 1) as WHISKEY,
        (SELECT category_name
        FROM sales
        WHERE category_name LIKE '%%VODKA%%' and total > 10000
        ORDER BY total DESC
        LIMIT 1) as VODKA
FROM sales
LIMIT 1
"""
pd.read_sql_query(sql, engine)

rum          whiskey           vodka
0  SPICED RUM  WHISKEY LIQUEUR  80 PROOF VODKA

In [15]:
# Which county sold the most amount of vodka during February 2014?
sql=""" 
SELECT county, SUM(total) as total
FROM sales
WHERE date BETWEEN '2014-2-01' AND '2014-2-28'
GROUP BY county
ORDER BY total DESC
LIMIT 1
"""
pd.read_sql_query(sql, engine)

county       total
0   Polk  8419114.08

In [34]:
# Which counties were in the top 10 counties for vodka sales in any month in 2014?
sql=""" 
SELECT DISTINCT county, total
FROM sales
WHERE category_name LIKE '%%VODKA%%' AND 
      date BETWEEN '2014-2-01' AND '2014-2-28'
GROUP BY county, total
ORDER BY total DESC
LIMIT 10
"""
pd.read_sql_query(sql, engine)

county     total
0  Pottawattamie  28782.00
1           Linn  19596.00
2           Polk  15676.80
3           Polk  14843.40
4           Polk  12541.44
5          Boone   9594.00
6        Carroll   9594.00
7           Polk   9594.00
8           Polk   8582.40
9          Scott   7838.40

In [9]:
# Create a report that shows how many times a county appeared in the “top 10 counties for vodka sales in a month” 
# list over the course of 2014.
sql=""" 
SELECT date, county, total
FROM sales
WHERE category_name LIKE '%%VODKA%%' AND 
      date BETWEEN '2014-01-01' AND '2014-12-31'
GROUP BY date, county, total
ORDER BY total DESC
LIMIT 10
"""
pd.read_sql_query(sql, engine)

date         county    total
0 2014-01-27  Pottawattamie  49470.0
1 2014-10-13         Dallas  39834.0
2 2014-11-07         Dallas  35988.0
3 2014-07-07         Dallas  34522.8
4 2014-06-23         Dallas  33843.6
5 2014-06-23         Dallas  31489.5
6 2014-05-29           Polk  29686.8
7 2014-06-23           Polk  29686.8
8 2014-02-10  Pottawattamie  28782.0
9 2014-06-23        Johnson  27180.0